lib/utils/fakeconn_test.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"io/ioutil"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+)
+
+// TestChConn validates that reads from the channel connection can be
+// canceled by setting a read deadline.
+func TestChConn(t *testing.T) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	t.Cleanup(func() { listener.Close() })
+
+	sshConnCh := make(chan sshConn)
+
+	go startSSHServer(t, listener, sshConnCh)
+
+	client, err := ssh.Dial("tcp", listener.Addr().String(), &ssh.ClientConfig{
```

Comment:
nit: I could imagine this racing with `startSSHServer` and hitting the `connection refused` if the server takes time to start. It is unlikely but not impossible unless this ensures that dialing only happens after the listener had started `Accept`ing.

Reply:
At least according to my experiments, it won't get connection refused since listener is already listening.